### PR TITLE
feat: support cache for batch_get in CachedMetaKvBackend

### PR DESCRIPTION
**Pull request by fengys1996**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

As title says: this pr mainly support cache for batch_get in CachedMetaKvBackend.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/catalog/src/kvbackend/client.rs [hunk 1]
```diff
@@ -154,8 +266,14 @@ impl KvBackend for CachedMetaKvBackend {
     async fn get(&self, key: &[u8]) -> Result<Option<KeyValue>> {
         let _timer = METRIC_CATALOG_KV_GET.start_timer();
 
+        let pre_version = Arc::new(Mutex::new(None));
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:269]: Add some comments, such as explaining that "invalidate_key" and "cache.try_get_with_by_ref" are not mutually exclusive.

File: src/meta-client/src/lib.rs [hunk 2]
```diff
@@ -36,6 +36,14 @@ pub struct MetaClientOptions {
     #[serde(with = "humantime_serde")]
     pub connect_timeout: Duration,
     pub tcp_nodelay: bool,
+    #[serde(default = "default_cached_max_capacity")]
+    pub cached_max_capacity: u64,
```
Comment (fengjiachun) [inline src/meta-client/src/lib.rs:40]:
```suggestion
    pub metadata_cache_max_capacity: u64,
```

**Comment by fengys1996** [comment]
@fengjiachun PTAL

File: src/catalog/src/kvbackend/client.rs [hunk 3]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
 use moka::future::{Cache, CacheBuilder};
 use snafu::{OptionExt, ResultExt};
 
-use crate::metrics::{METRIC_CATALOG_KV_GET, METRIC_CATALOG_KV_REMOTE_GET};
+use crate::metrics::{
+    METRIC_CATALOG_KV_BATCH_GET, METRIC_CATALOG_KV_GET, METRIC_CATALOG_KV_REMOTE_GET,
+};
+
+const DEFAULT_MAX_CACHE_CAPACITY: u64 = 10000;
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:42]:
```suggestion
const DEFAULT_CACHE_MAX_CAPACITY: u64 = 10000;
```

Comment (fengjiachun) [reply src/catalog/src/kvbackend/client.rs:42]: Keep the same as : https://github.com/GreptimeTeam/greptimedb/pull/3277/files#diff-8c8314e72e06586a9aea0e9e4a2ed8c509312768340be34a6204101483e517e6R73

File: src/catalog/src/kvbackend/client.rs [hunk 4]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
...
+const DEFAULT_CACHE_TTL: Duration = Duration::from_secs(10 * 60);
+const DEFAULT_CACHE_TTI: Duration = Duration::from_secs(5 * 60);
+
+pub struct CachedMetaKvBackendBuilder {
+    max_cache_capacity: Option<u64>,
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:47]:
```suggestion
    cache_max_capacity: Option<u64>,
```

File: src/catalog/src/kvbackend/client.rs [hunk 5]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
...
+            version,
+        }
+    }
+}
 
 pub type CacheBackendRef = Arc<Cache<Vec<u8>, KeyValue>>;
 
+/// A wrapper of `MetaKvBackend` with cache support.
+///
+/// CachedMetaKvBackend is mainly used to read MetaData information from MetaSrv, and provides
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:111]:
```suggestion
/// CachedMetaKvBackend is mainly used to read metadata information from Metasrv, and provides
```

File: src/catalog/src/kvbackend/client.rs [hunk 6]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
...
+/// cache for get and batch_get. One way to trigger cache invalidation of CachedMetaKvBackend:
+/// when Metadata information changes, Metasrv will broadcast a MetaData invalidation request.
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:113]:
```suggestion
/// when metadata information changes, Metasrv will broadcast a metadata invalidation request.
```

File: src/catalog/src/kvbackend/client.rs [hunk 7]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
...
+///
+/// Therefore, it is recommended to use CachedMetaKvBackend to only read MetaData related
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:115]:
```suggestion
/// Therefore, it is recommended to use CachedMetaKvBackend to only read metadata related
```

File: src/catalog/src/kvbackend/client.rs [hunk 8]
```diff
@@ -175,29 +287,40 @@ impl KvBackend for CachedMetaKvBackend {
         }
         .map_err(|e| GetKvCache {
             err_msg: e.to_string(),
-        })
+        });
+
+        // "invalidate_key" and "cache.try_get_with_by_ref" are not mutually exclusive. So we need
```
Comment (fengjiachun) [inline src/catalog/src/kvbackend/client.rs:292]:
```suggestion
        // "cache.invalidate_key" and "cache.try_get_with_by_ref" are not mutually exclusive. So we need
```

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/catalog/src/kvbackend/client.rs [hunk 9]
```diff
@@ -96,7 +171,38 @@ impl KvBackend for CachedMetaKvBackend {
     }
 
     async fn batch_get(&self, req: BatchGetRequest) -> Result<BatchGetResponse> {
-        self.kv_backend.batch_get(req).await
+        let _timer = METRIC_CATALOG_KV_BATCH_GET.start_timer();
+
+        let mut kvs = Vec::with_capacity(req.keys.len());
+        let mut unhit_keys = Vec::with_capacity(req.keys.len());
```
Comment (niebayes) [inline src/catalog/src/kvbackend/client.rs:177]: non-blocking:
```suggestion
        let mut miss_keys = Vec::with_capacity(req.keys.len());
```

File: src/catalog/src/kvbackend/client.rs [hunk 10]
```diff
@@ -33,18 +35,91 @@ use meta_client::client::MetaClient;
...
+            cache_ttl: None,
+            cache_tti: None,
+            meta_client,
+        }
+    }
+
+    pub fn cache_max_capacity(mut self, cache_max_capacity: u64) -> Self {
+        self.cache_max_capacity.replace(cache_max_capacity);
+        self
+    }
```
[the remark this answers is not in the document]
Comment (fengys1996) [reply src/catalog/src/kvbackend/client.rs:66]: I think both writing methods are fine, but the following one may cost more to build Self. :smiley:

**Comment by fengys1996** [comment]
@niebayes PTAL

**Review by niebayes — APPROVED** [review]

**Pull request merged**